proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+          )
+      expected :: IdCard
+      expected = MkIdCard "John Doe" (MkAddress 42 "Elm Street" "Springwood" 12345)
+  assertEqual "John Doe ID is the same in both building schemes" expected actual
+```
+
+The syntax looks quite verbose compared to naive structure building, because the API is designed to allow efficient structure building, little-by-little, from dynamic data, instead of building the whole structure at once from static data (where it is no more efficient than a copy using existing `compactAdd#`).
+
+The API uses linear types to ensure (at compile-time) that every field of the data structure is wrote to exactly once before the structure can be read. More details on the API can be found in this [JFLA 2024 article](https://inria.hal.science/hal-04406360/document) and [this blog post](https://www.tweag.io/blog/2024-03-07-dps-haskell/).
+
+### Library side: more performant S-expression parser
+
+[I implemented a S-expression parser]((https://github.com/tweag/linear-base/blob/93dd7769d86a4c918934be110151faf382a35f38/examples-version-changes/ghc-dps-compact/after/Compact/SExpr.hs)) in two different fashions: the first one `parseWithoutDest` deserializes into the usual GC heap, and the second `parseWithDest` uses this proposal's primitives (via the [aforementioned safe library-side API](https://github.com/tweag/linear-base/pull/450)) to directly deserialize data into a compact region. Here's the benchmark for a 4MB S-expression file:
+
+```text
+parseWithoutDest (parse into GC heap, stay in GC heap ***): OK
+  358  ms ± 6.8 ms, 196 MB allocated, 402 MB copied, 441 MB peak memory
+
+parseWithoutDest (parse into GC heap, then copy into compact region): OK
+  332  ms ± 4.4 ms, 327 MB allocated, 362 MB copied, 377 MB peak memory
+
+parseWithDest (parse directly into compact region): OK
+  136  ms ± 5.4 ms, 283 MB allocated,  37 MB copied, 269 MB peak memory
+```
+
+(`***`_The  program only parses the document and then shutdowns, so in a real-world use case doing something non trivial with the parsed data, the time spent by the GC traversing the parsed dataset again and again in `parseWithoutDest (parse into GC heap, stay in GC heap` would be significantly more important._)
+
+Unlike the previous example, here the code for the parser is very similar and equally verbose for both the naive and DPS version, so there isn't much extra effort needed to get these performance benefits.
+
+## Effect and Interactions
+
+Primitive `compactAddHollow#` allow for direct allocation in a compact region, without copy, as stated in the [Motivation](#motivation) section. It needs to be combined with `reifyInfoTablePtr#` to properly allocate requested constructor objects with the right info table pointer, because as a RTS primitive, `compactAddHollow#` doesn't have access itself to the info table pointer to set for the constructor (all information about info tables has long been inlined/erased when the primitive is actually executed).
```

Comment:
```suggestion
The `compactAddHollow#` primitive allows for direct allocation in a compact region, without copy, as stated in the [Motivation](#motivation) section. It needs to be combined with `reifyInfoTablePtr#` to properly allocate requested constructor objects with the right info table pointer, because as a RTS primitive, `compactAddHollow#` doesn't have access itself to the info table pointer to set for the constructor (all information about info tables has long been inlined/erased when the primitive is actually executed).
```